File: src/main/auth/device-flow.ts
export const GITHUB_SCOPES = 'repo read:org'

const DEVICE_CODE_URL = 'https://github.com/login/device/code'
const ACCESS_TOKEN_URL = 'https://github.com/login/oauth/access_token'

export interface DeviceCodeInfo {
  userCode: string
  verificationUri: string
  deviceCode: string
  /** Seconds GitHub asks us to wait between polls. */
  interval: number
  expiresIn: number
}

export interface PollDeps {
  fetchFn?: typeof fetch
  sleep?: (ms: number) => Promise<void>
}

interface DeviceCodeResponse {
  device_code?: string
  user_code?: string
  verification_uri?: string
  interval?: number
  expires_in?: number
  error?: string
  error_description?: string
}

interface TokenResponse {
  access_token?: string
  error?: string
  error_description?: string
}

async function postJson<T>(
  url: string,
  body: Record<string, string>,
  fetchFn: typeof fetch,
): Promise<T> {
  const response = await fetchFn(url, {
    method: 'POST',
    headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  if (!response.ok) {
    throw new Error(`GitHub returned ${response.status} from ${url}`)
  }
  return (await response.json()) as T
}

function describeError(error: string, description?: string): string {
  return description ? `${error}: ${description}` : error
}

export async function requestDeviceCode(
  clientId: string,
  fetchFn: typeof fetch = fetch,
): Promise<DeviceCodeInfo> {
  const data = await postJson<DeviceCodeResponse>(
    DEVICE_CODE_URL,
    { client_id: clientId, scope: GITHUB_SCOPES },
    fetchFn,
  )

  if (data.error !== undefined) {
    throw new Error(describeError(data.error, data.error_description))
  }
  if (!data.device_code || !data.user_code || !data.verification_uri) {
    throw new Error('GitHub sent back an incomplete device code response')
  }

  return {
    userCode: data.user_code,
    verificationUri: data.verification_uri,
    deviceCode: data.device_code,
    interval: data.interval ?? 5,
    expiresIn: data.expires_in ?? 900,
  }
}

const defaultSleep = (ms: number): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, ms))

/**
 * Polls until the user approves the device in their browser. Resolves with the
 * access token, or rejects if the user denies it or the code expires.
 */
export async function pollForToken(
  clientId: string,
  info: DeviceCodeInfo,
  deps: PollDeps = {},
): Promise<string> {
  const fetchFn = deps.fetchFn ?? fetch
  const sleep = deps.sleep ?? defaultSleep
  let intervalMs = info.interval * 1000
  let elapsedMs = 0
  const expiresInMs = info.expiresIn * 1000

  for (;;) {
    await sleep(intervalMs)
    elapsedMs += intervalMs

    if (elapsedMs >= expiresInMs) {
      throw new Error('That code expired. Sign in again.')
    }

    const data = await postJson<TokenResponse>(
      ACCESS_TOKEN_URL,
      {
        client_id: clientId,
        device_code: info.deviceCode,
        grant_type: 'urn:ietf:params:oauth:grant-type:device_code',
      },
      fetchFn,
    )

    if (data.access_token !== undefined) return data.access_token

    switch (data.error) {
      case 'authorization_pending':
        break
      case 'slow_down':
        intervalMs += 5000
        break
      default:
        throw new Error(describeError(data.error ?? 'unknown_error', data.error_description))
    }
  }
}

File: Sources/PulloverKit/Auth/DeviceFlow.swift
import Foundation

/// `repo` and `read:org` are the narrowest scopes that can still see pull
/// requests in private repositories and review requests that arrived through a team.
public let githubScopes = "repo read:org"

public struct DeviceCode: Sendable, Equatable {
    public var userCode: String
    public var verificationURI: String
    public var deviceCode: String
    /// Seconds GitHub asks us to wait between polls.
    public var interval: Int
    public var expiresIn: Int
}

public enum DeviceFlowError: Error, LocalizedError, Equatable {
    case github(String)
    case expired
    case incomplete
    case http(Int)

    public var errorDescription: String? {
        switch self {
        case let .github(message): message
        case .expired: "That code expired. Sign in again."
        case .incomplete: "GitHub sent back an incomplete device code response"
        case let .http(status): "GitHub returned \(status) during sign-in"
        }
    }
}

/// Posts a form as JSON and returns the JSON response body.
public protocol HTTPPoster: Sendable {
    func postJSON(_ url: URL, body: [String: String]) async throws -> Data
}

public struct URLSessionPoster: HTTPPoster {
    public init() {}

    public func postJSON(_ url: URL, body: [String: String]) async throws -> Data {
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(body)
        let (data, response) = try await URLSession.shared.data(for: request)
        if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            throw DeviceFlowError.http(http.statusCode)
        }
        return data
    }
}

/// GitHub's OAuth Device Flow: no redirect URI, no client secret — the user
/// types a short code into github.com and the app polls until they approve.
public struct DeviceFlow: Sendable {
    static let deviceCodeURL = URL(string: "https://github.com/login/device/code")!
    static let accessTokenURL = URL(string: "https://github.com/login/oauth/access_token")!

    private struct CodeResponse: Decodable {
        var device_code: String?
        var user_code: String?
        var verification_uri: String?
        var interval: Int?
        var expires_in: Int?
        var error: String?
        var error_description: String?
    }

    private struct TokenResponse: Decodable {
        var access_token: String?
        var error: String?
        var error_description: String?
    }

    public let clientID: String
    private let poster: any HTTPPoster
    private let sleep: @Sendable (Duration) async throws -> Void

    public init(
        clientID: String,
        poster: any HTTPPoster = URLSessionPoster(),
        sleep: @escaping @Sendable (Duration) async throws -> Void = { try await Task.sleep(for: $0) }
    ) {
        self.clientID = clientID
        self.poster = poster
        self.sleep = sleep
    }

    private static func describe(_ error: String, _ description: String?) -> String {
        description.map { "\(error): \($0)" } ?? error
    }

    public func requestCode() async throws -> DeviceCode {
        let data = try await poster.postJSON(Self.deviceCodeURL, body: ["client_id": clientID, "scope": githubScopes])
        let response = try JSONDecoder().decode(CodeResponse.self, from: data)
        if let error = response.error {
            throw DeviceFlowError.github(Self.describe(error, response.error_description))
        }
        guard let deviceCode = response.device_code, let userCode = response.user_code,
              let uri = response.verification_uri else { throw DeviceFlowError.incomplete }
        return DeviceCode(
            userCode: userCode,
            verificationURI: uri,
            deviceCode: deviceCode,
            interval: response.interval ?? 5,
            expiresIn: response.expires_in ?? 900
        )
    }

    /// Polls until the user approves the device in their browser. Returns the
    /// access token, or throws if they deny it or the code expires.
    public func pollForToken(_ code: DeviceCode) async throws -> String {
        var interval = code.interval
        var elapsed = 0

        while true {
            try await sleep(.seconds(interval))
            elapsed += interval
            if elapsed >= code.expiresIn { throw DeviceFlowError.expired }

            let data = try await poster.postJSON(Self.accessTokenURL, body: [
                "client_id": clientID,
                "device_code": code.deviceCode,
                "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
            ])
            let response = try JSONDecoder().decode(TokenResponse.self, from: data)
            if let token = response.access_token { return token }

            switch response.error {
            case "authorization_pending": continue
            case "slow_down": interval += 5
            default: throw DeviceFlowError.github(Self.describe(response.error ?? "unknown_error", response.error_description))
            }
        }
    }
}
